unit-tests/src/test/java/org/eclipse/collections/impl/utility/LazyIterateTest.java

```diff
@@ -179,4 +184,77 @@ public void classIsNonInstantiable()
+        MutableList<Integer> iterable2 = Lists.mutable.with(2, 3, 4);
+        MutableBag<Pair<Integer, Integer>> expectedCartesianProduct1 = Bags.mutable.with(
+                Tuples.pair(1, 2),
+                Tuples.pair(2, 2),
+                Tuples.pair(1, 3),
+                Tuples.pair(2, 3),
+                Tuples.pair(1, 4),
+                Tuples.pair(2, 4));
+        Assert.assertEquals(expectedCartesianProduct1, LazyIterate.cartesianProduct(iterable1, iterable2).toBag());
+        MutableBag<Pair<Integer, Integer>> expectedCartesianProduct2 = Bags.mutable.with(
+                Tuples.pair(2, 1),
+                Tuples.pair(3, 1),
+                Tuples.pair(4, 1),
+                Tuples.pair(2, 2),
+                Tuples.pair(3, 2),
+                Tuples.pair(4, 2));
+        Assert.assertEquals(expectedCartesianProduct2, LazyIterate.cartesianProduct(iterable2, iterable1).toBag());
+    }
+
+    @Test
+    public void cartesianProductSameElements()
+    {
+        MutableList<Integer> iterable1 = Lists.mutable.with(1, 2);
+        MutableList<Integer> iterable2 = Lists.mutable.with(1, 2);
+        MutableBag<Pair<Integer, Integer>> expectedCartesianProduct = Bags.mutable.with(
+                Tuples.pair(1, 1),
+                Tuples.pair(1, 2),
+                Tuples.pair(2, 1),
+                Tuples.pair(2, 2));
+        Assert.assertEquals(expectedCartesianProduct, LazyIterate.cartesianProduct(iterable1, iterable2).toBag());
+    }
+
+    @Test
+    public void cartesianProductWithFunction()
+    {
+        MutableList<Integer> iterable1 = Lists.mutable.with(1, 2);
+        MutableList<Integer> iterable2 = Lists.mutable.with(2, 3, 4);
+        MutableBag<Pair<Integer, Integer>> expectedCartesianProduct = Bags.mutable.with(
+                Tuples.pair(1, 2),
+                Tuples.pair(2, 2),
+                Tuples.pair(1, 3),
+                Tuples.pair(2, 3),
+                Tuples.pair(1, 4),
+                Tuples.pair(2, 4));
+        Assert.assertEquals(
+                expectedCartesianProduct,
+                LazyIterate.cartesianProduct(iterable1, iterable2, Tuples::pair).toBag());
+        MutableBag<MutableList<Integer>> expectedCartesianProduct2 = Bags.mutable.with(
+                Lists.mutable.with(2, 1),
+                Lists.mutable.with(3, 1),
+                Lists.mutable.with(4, 1),
+                Lists.mutable.with(2, 2),
+                Lists.mutable.with(3, 2),
+                Lists.mutable.with(4, 2));
+        Assert.assertEquals(
+                expectedCartesianProduct2,
+                LazyIterate.cartesianProduct(iterable2, iterable1, Lists.mutable::with).toBag());
+    }
+
+    @Test
+    public void cartesianProduct_empty()
+    {
+        Assert.assertEquals(
+                Bags.mutable.empty(),
+                LazyIterate.cartesianProduct(
+                        Lists.mutable.with(1, 2),
+                        Lists.mutable.empty()).toBag());
+    }
```

Comment:
Please test with duplicates. The reason this functionality only ever existed for sets is because it's easier to answer how it ought to behave in the case of duplicates (there aren't any). What should the cartesian product be between (1, 1) and ("same", "same")?

Reply:
hmm, very good question :) 

I would say that the result should be `((1, "same"), (1, "same"), (1, "same"), (1, "same"))` because I'm trusting the two collections to handle duplicates or not.

So the behaviour should be different depending on the collections given to the method and if you need no duplicates, then you should use sets... but maybe that's not compatible with the philosophy of EC? WDYT? @donraab @motlin

Reply:
As long as the behavior is true to the resulting collection in my opinion it’s fine. For lazy, we can make this to be non terminating and the resultant collection will take care of collection semantics.

Reply:
As long as the behavior is true to the resulting collection in my opinion it’s fine. For lazy, we can make this to be non terminating and the resultant collection will take care of collection semantics.

Reply:
@victornoel Please write some tests with duplicates and outputs of Bag, Set, List.  Then it will be easier for us to see what the resulting behavior looks like.  It would be good to document our resulting agreement in the JavaDoc for the method.

Reply:
@donraab yep, good idea, I will do this!

Reply:
@donraab done

Reply:
@victornoel Thanks!